File: src/main/java/de/tectoast/emolga/commands/GuildCommandEvent.java
package de.tectoast.emolga.commands;

import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.TextChannel;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.events.message.MessageReceivedEvent;

import static de.tectoast.emolga.utils.Constants.FLOID;
import static de.tectoast.emolga.utils.Constants.MYTAG;

public class GuildCommandEvent extends GenericCommandEvent {
    private final Member member;
    private final Guild guild;
    private final TextChannel tco;
    private final String usedName;
    private Command.ArgumentManager manager;

    public GuildCommandEvent(Command c, MessageReceivedEvent e) throws Exception {
        super(e.getMessage());
        this.member = e.getMember();
        this.guild = e.getGuild();
        this.tco = e.getTextChannel();
        Command.ArgumentManagerTemplate template = c.getArgumentTemplate();
        if (template != null)
            this.manager = template.construct(e, c);
        this.usedName = Command.WHITESPACES_SPLITTER.split(getMsg())[0].substring(c.getPrefix().length());
        new Thread(() -> {
            try {
                (manager != null ? manager.executor() : c).process(this);
            } catch (Exception ex) {
                ex.printStackTrace();
                tco.sendMessage("Es ist ein Fehler beim Ausführen des Commands aufgetreten!\nWenn du denkst, dass dies ein interner Fehler beim Bot ist, melde dich bitte bei Flo (%s).\n".formatted(MYTAG) + c.getHelp(e.getGuild()) + (member.getIdLong() == FLOID ? "\nJa Flo, du sollst dich auch bei ihm melden du Kek! :^)" : "")).queue();
            }
        }, "CMD " + c.getName()).start();
    }

    public GuildCommandEvent(Command c, SlashCommandInteractionEvent e) throws Exception {
        super(e);
        this.member = e.getMember();
        this.guild = e.getGuild();
        this.tco = e.getTextChannel();
        this.usedName = e.getName();
        Command.ArgumentManagerTemplate template = c.getArgumentTemplate();
        if (template != null)
            this.manager = template.construct(e, c);
        new Thread(() -> {
            try {
                (manager != null ? manager.executor() : c).process(this);
            } catch (Exception ex) {
                ex.printStackTrace();
                tco.sendMessage("Es ist ein Fehler beim Ausführen des Commands aufgetreten!\nWenn du denkst, dass dies ein interner Fehler beim Bot ist, melde dich bitte bei Flo (%s).\n".formatted(MYTAG) + c.getHelp(e.getGuild()) + (member.getIdLong() == FLOID ? "\nJa Flo, du sollst dich auch bei ihm melden du Kek! :^)" : "")).queue();
            }
        }, "CMD " + c.getName()).start();
    }


    public String getUsedName() {
        return usedName;
    }

    public Member getMember() {
        return member;
    }

    public Guild getGuild() {
        return guild;
    }

    public Command.ArgumentManager getArguments() {
        return manager;
    }

    public TextChannel getChannel() {
        return tco;
    }

    public void deleteMessage() {
        this.getMessage().delete().queue();
    }
}

File: src/main/kotlin/de/tectoast/emolga/commands/GuildCommandEvent.kt
package de.tectoast.emolga.commands

import de.tectoast.emolga.commands.Command.ArgumentManager
import de.tectoast.emolga.utils.Constants
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.channel.middleman.GuildMessageChannel
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.message.MessageReceivedEvent

class GuildCommandEvent : GenericCommandEvent {
    val member: Member
    val guild: Guild
    val textChannel: GuildMessageChannel
    val usedName: String
    val command: Command
    lateinit var arguments: ArgumentManager
        private set


    constructor(c: Command, e: MessageReceivedEvent) : super(e.message) {
        member = e.member!!
        guild = e.guild
        this.textChannel = e.guildChannel
        command = c
        usedName = Command.WHITESPACES_SPLITTER.split(msg!!)[0].substring(c.prefix.length)
    }

    suspend fun initialize(c: Command, e: MessageReceivedEvent) {
        arguments = c.argumentTemplate.construct(e, c)
    }

    suspend fun execute() {
        coroutineScope {
            this.launch {
                try {
                    arguments.executor.process(this@GuildCommandEvent)
                } catch (ex: Exception) {
                    ex.printStackTrace()
                    send(
                        "Es ist ein Fehler beim Ausführen des Commands aufgetreten!\nWenn du denkst, dass dies ein interner Fehler beim Bot ist, melde dich bitte bei Flo (${Constants.MYTAG})."
                            .condAppend(
                                member.idLong == Constants.FLOID,
                                "\nJa Flo, du sollst dich auch bei ihm melden du Kek :^)"
                            )
                    )
                    Command.sendToMe("Error " + textChannel.asMention + " Server: " + textChannel.guild.name + " Msg: " + msg + " View Console")
                }
            }
        }
    }

    private fun send(msg: String) {
        slashCommandEvent?.reply(msg)?.queue() ?: textChannel.sendMessage(msg).queue()
    }

    constructor(c: Command, e: SlashCommandInteractionEvent) : super(e) {
        member = e.member!!
        guild = e.guild!!
        command = c
        this.textChannel = e.guildChannel
        usedName = e.name
    }

    suspend fun initialize(c: Command, e: SlashCommandInteractionEvent) {
        arguments = c.argumentTemplate.construct(e, c)
    }

}
